Move transport version constants to TransportVersions

rjernst opened the pull request:
Constants for TransportVersion currently live alongeside the class definition. This has been fine since there was only one set of constants. However, to support serverless, some constants will need to be defined elsewhere.

This commit moves the existing constants to a new holder class, TransportVersions. It is almost entirely mechanical, using IntelliJ move members. The only non mechanical part was slightly shifting how CURRENT is found, defining a LATEST in TransportVersions that is automatically calculated (since we already have it, no need to manually define it).

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

thecoop reviewed (approved):
It's worth letting ES devs know that how transport versions will be managed is changing in staged

thecoop commented on server/src/main/java/org/elasticsearch/TransportVersion.java:
```diff
@@ -65,191 +55,27 @@
             var versionExtension = VersionExtension.load();
             if (versionExtension == null) {
-                return fallback;
+                return TransportVersions.LATEST;
```
We need some clear disambiguation between `TransportVersion.CURRENT` and `TransportVersions.LATEST` in the docs

rjernst replied:
I renamed LATEST to LATEST_DEFINED, and added a comment. This is just an internal mechanism between these to classes to find the fallback version if current hasn't been plugged in.

thecoop commented on server/src/main/java/org/elasticsearch/TransportVersion.java:
```diff
@@ -279,7 +105,7 @@ public static TransportVersion max(TransportVersion version1, TransportVersion v
      */
     public static boolean isCompatible(TransportVersion version) {
-        return version.onOrAfter(MINIMUM_COMPATIBLE);
+        return version.onOrAfter(TransportVersions.MINIMUM_COMPATIBLE);
```
Maybe MINIMUM_COMPATIBLE and MINIMUM_CCS_VERSION should continue to be defined here, alongside `current()`, to prevent this cross-referencing?

rjernst replied:
There would still be cross referencing, since the values of MINIMUM_COMPATIBLE and MINIMUM_CCS_VERSION are defined statically. I think current is a bit special, while those constants are fixed (at least for now). If we make them adjustable in the future the way current is, then we could move them back as methods.

thecoop commented:
We also need to update the release automation and the docs in `CONTRIBUTING.md`

rjernst committed:
Constants for TransportVersion currently live alongeside the class definition. This has been fine since there was only one set of constants. However, to support serverless, some constants will need to be defined elsewhere.

This commit moves the existing constants to a new holder class, TransportVersions. It is almost entirely mechanical, using IntelliJ move members. The only non mechanical part was slightly shifting how CURRENT is found, defining a LATEST in TransportVersions that is automatically calculated (since we already have it, no need to manually define it).

rjernst commented:
> We also need to update the release automation and the docs in CONTRIBUTING.md

I'll leave that to a followup, since this PR is large and I want it to remain as mechanical as possible. It is difficult to keep up to date.

rjernst committed:
spotless

rjernst committed:
Merge branch 'main' into transport_versions/move

rjernst committed:
updates

rjernst committed:
fix oops

rjernst committed:
Merge branch 'main' into transport_versions/move

rjernst committed:
fix compile

rjernst committed:
spotless

rjernst committed:
fix line lengths

rjernst committed:
Merge branch 'main' into transport_versions/move

rjernst committed:
fixes

rjernst committed:
Merge branch 'main' into transport_versions/move

rjernst committed:
spotless

rjernst committed:
Merge branch 'main' into transport_versions/move

rjernst committed:
fix compile
